### PR TITLE
add complex-type support to avro-to-pinot schema inference

### PR DESCRIPTION
**Pull request by yupeng9**
## Description
Part of https://github.com/apache/incubator-pinot/issues/6904

Add complex-type support to avro-to-pinot schema inference. Now users can use the command like the following to infer Pinot schema from Avro schema with complex-structure

```
AvroSchemaToPinotSchema -timeColumnName fields.hoursSinceEpoch -avroSchemaFile  pinot-plugins/pinot-input-format/pinot-avro-base/src/test/resources/fake_avro_nested_schema.avsc -pinotSchemaName schema -outputDir /tmp/test -unnestFields=entries 
```

### CODE REVIEW

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroUtils.java [hunk 1]
```diff
@@ -277,4 +280,86 @@ public static DataType extractFieldDataType(Field field) {
       return fieldSchema;
     }
   }
+
+  private static void extractSchemaWithComplexTypeHandling(org.apache.avro.Schema fieldSchema,
+      List<String> unnestFields, String delimiter, String path, Schema pinotSchema,
+      @Nullable Map<String, FieldSpec.FieldType> fieldTypeMap, @Nullable TimeUnit timeUnit) {
+    org.apache.avro.Schema.Type fieldType = fieldSchema.getType();
+    if (fieldType == org.apache.avro.Schema.Type.UNION) {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroUtils.java:288]: Consider using switch?

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroUtils.java [hunk 2]
```diff
@@ -277,4 +280,86 @@ public static DataType extractFieldDataType(Field field) {
...
+              new DateTimeFormatSpec(1, timeUnit.toString(), DateTimeFieldSpec.TimeFormat.EPOCH.toString()).getFormat(),
+              new DateTimeGranularitySpec(1, timeUnit).getGranularity()));
+          break;
+        default:
+          throw new UnsupportedOperationException("Unsupported field type: " + fieldType + " for field: " + name);
+      }
+    }
+  }
+
+  private static String concat(String delimiter, String path, String component) {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroUtils.java:362]: Directly inline this method might be more readable?

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java [hunk 3]
```diff
@@ -52,6 +52,25 @@ public static DataType valueOf(Schema.Type avroType) {
...
+  public static boolean isPrimitiveType(Schema.Type avroType) {
+    switch (avroType) {
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+      case BOOLEAN:
+      case STRING:
+      case ENUM:
+      case BYTES:
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java:67]: Currently Pinot does not support MV BYTES, so not sure if we want to count it as primitive type

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

**Comment by yupeng9** [comment]
The integration test failure not related to this PR.

**Pull request merged**